src/main/java/org/folio/rest/impl/LoginAPI.java

```diff
@@ -288,184 +268,255 @@ public void postAuthnLogin(String userAgent, String xForwardedFor,
-          ));
+      handleCrossTenantLogin(entity, tenantId, okapiHeaders, asyncResultHandler)
+        .onComplete(ar -> {
+          if (ar.failed()) {
+            // asyncResultHandler was updated in handleCrossTenantLogin()
+            return;
+          }
+          String newTenantId = ar.result();
+          Future<JsonObject> userVerified;
+          if (entity.getUserId() != null && !requireActiveUser) {
+            logger.debug("No need to look up user id");
+            userVerified = Future.succeededFuture(new JsonObject()
+                .put("id", entity.getUserId()).put(ACTIVE, true)
+                .put(USERNAME, "__undefined__"));
+          } else {
+            logger.debug("Need to look up user id");
+            if (entity.getUserId() != null) {
+              userVerified = lookupUser(null, entity.getUserId(), newTenantId, okapiURL, requestToken);
+            } else {
+              userVerified = lookupUser(entity.getUsername(), null, newTenantId, okapiURL, requestToken);
+            }
+          }
+          userVerified.onComplete(verifyResult -> loginAfterUserVerified(verifyResult, entity, newTenantId, okapiURL,
+              requestToken, userAgent, xForwardedFor, okapiHeaders, asyncResultHandler, vertxContext));
+        });
+    } catch(Exception e) {
+      logger.debug("Error running on verticle for postAuthnLogin: " + e.getLocalizedMessage());
+      asyncResultHandler.handle(Future.succeededFuture(PostAuthnLoginResponse.respond500WithTextPlain(INTERNAL_ERROR)));
+    }
+  }
 
-        } else {
-          //User's okay, let's try to login
-          try {
-            JsonObject userObject = verifyResult.result();
-            if (!userObject.containsKey("id")) {
-              logger.error("No 'id' key in returned user object");
-              asyncResultHandler.handle(Future.succeededFuture(
-                  PostAuthnLoginResponse.respond500WithTextPlain(
-                      "No user id could be found")));
+  private Future<String> handleCrossTenantLogin(LoginCredentials credentials, String tenantId,
+                                                Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler) {
+    if (credentials.getTenantId() == null || credentials.getTenantId().equals(okapiHeaders.get(XOkapiHeaders.TENANT)))
+      return Future.succeededFuture(tenantId);
+    Promise<String> promise = Promise.promise();
+    userService.getUserTenants(tenantId, credentials.getUsername(), credentials.getUserId(), credentials.getTenantId(),
+      LoginConfigUtils.encodeJsonHeaders(okapiHeaders), ar -> {
+        if (ar.failed()) {
+          promise.fail(ar.cause());
+          return;
+        }
+        try {
+          List<UserTenant> userTenants = ar.result().stream()
+            .map(obj -> ((JsonObject)obj).mapTo(UserTenant.class))
+            .collect(toList());
+          if (userTenants.isEmpty()) {
```

Comment:
Please add some comments in sections of if statement to summarize the logic. The three main sections are userTenants is empty, userTenants has more than one & the last case.

Reply:
I will add comments. The last case is for a single matching user-tenant association.